README.md
# quandico

### Abstract

The tool `quandico` applies statistical methods to detect [copy number 
variations][cnv] (CNV) in a sample by comparing the 
read counts from next generation sequencing (NGS) performed after PCR-enrichment of regions of interest, typically a 
set of genes with known or expected relevance for the sample, e.g. genes that play a role in cancer. Counts from a normal 
control (ideally matched normal from the same individual, e.g. healthy tissue) are required. 

A detailed description can be found in the publication of this tool: **Quantitative analysis of differences in copy numbers using read depth obtained from PCR-enriched samples and controls**: [BMC Bioinformatics 2015, 16:17](http://www.biomedcentral.com/1471-2105/16/17 "BMC Bioinformatics"). 

## Implementation
The main script is written in `R` and there are two helper scripts and one main driver script written in Perl.

## Workflow
1. PCR-enrichment of target regions, sequencing and read-mapping (SAM/BAM file)
2. Extract counts for every PCR amplicon (`qgetcounts`)
3. Cluster read-counts into regions (`qcluster`)
4. Compare counts per cluster and perform CNV calling (`quandico`)

Steps 2, 3 and 4 can be performed with one single command line call of the Perl 
driver script `quandico`. Step 4 alone can be performed inside `R`.

## Requirements
Running `quandico` requires `R` with some commonly available packages from [CRAN][cran]. 
Please visit [The `R`-project homepage][r] for advice how to install `R` on your system.
Dependencies for `quandico` should be installed automatically when installing the package.

Accessory applications such as a command-line driver script (`quandico`), the script to extract counts of 
mapped reads (`qgetcounts`) and the region clustering script (`qcluster`) require Perl. Perl is already installed on 
most Linux systems, please check [Perl.org][perl] for details. If you are using Windows we 
recommend [Strawberry Perl][strawberry]. All dependencies should automatically be installed 
from [CPAN][cpan].

## Installation

### Fast

Download the [`install.sh`][installer] shell script for Unix/Linux. 
Check the configuration in the first section of that script, adjust 
parameters and then start it in a clean directory, e.g.:

```bash
$ mkdir quandico
$ cd quandico
$ wget https://github.com/reineckef/quandico/raw/master/install.sh
$ sh install.sh
```

This is not well tested and may fail on various systems for various 
reasons.

### Detailed

Install these `R` packages, either by using `R> install.packages('name')` inside an `R` session 
or by running the install command from a command line: `$ R CMD INSTALL 'name'`:

 * MASS
 * ggplot2
 * grid
 * gridExtra
 * naturalsort
 * scales

The string *n.m* will be used for the version number (major.minor) of `quandico`. This was **1.14** by the time of writing.

 * Download [the latest release][latest] and install the packaged `R` code of `quandico` (file `quandico_n.m.tar.gz`):

`$ R CMD INSTALL quandico_n.m.tar.gz`

**_optional_ but _recommended_**

 * Download [the latest release][latest] and install the Perl module `QUANDICO` (file `QUANDICO-vn.m.tar.gz`) 
 using `make` on Linux, and `dmake` on Windows:

```bash
$ tar xvfz QUANDICO-vn.m.tar.gz
$ cd QUANDICO-vn.m
$ perl Makefile.PL
$ [d]make
$ [d]make test
$ [d]make install # run this as sudo/root for system install
```

Alternatively, you can use `cpanm` from `App::cpanminus` to install to module directly from the downloaded archive. 
We recommend to use `--verbose` mode:
	
`$ cpanm --verbose QUANDICO-vn.m.tar.gz` # run this as sudo/root for system install

To start from mapped reads in SAM/BAM format, the Perl script `qgetcounts` will call `samtools` (version 1.1 or later) to 
extract the counts. Therefore, `samtools` needs to be installed, please visit [Samtools][samtools] for advice.

## Example Data

Example data files to test the tools are available on Google drive. To start from mapped reads (in SAM or BAM files, please 
note that this step additionally requires **samtools** version >= 1.1).

* [CNA902Y.bed](https://drive.google.com/open?id=0BzLnl09R3GITQjBUZUFVcy1BNFk&authuser=0) amplicon coordinates
* [M62_NA13019.bam](https://drive.google.com/open?id=0BzLnl09R3GITMzNyakhveTh3UVE&authuser=0) example case (sample) **:warning: 252 MB !**
* [M62_NA12878.bam](https://drive.google.com/open?id=0BzLnl09R3GITSnU1TlVRSjRXRHM&authuser=0) example control (reference) **:warning: 222 MB !**

Please note, these files can also be downloaded using wget (sometimes these links do not work, for unclear reason):

```bash
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITQjBUZUFVcy1BNFk -O CNA902Y.bed
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITMzNyakhveTh3UVE -O M62_NA13019.bam
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITSnU1TlVRSjRXRHM -O M62_NA12878.bam
```

To skip count extraction and start with clustering, the extracted counts of these samples are also available:

* [M62_NA13019.tsv](https://drive.google.com/open?id=0BzLnl09R3GITUDZ0aXFBd2pDR0k&authuser=0) control counts (sample)
* [M62_NA12878.tsv](https://drive.google.com/open?id=0BzLnl09R3GITWU9xTndtZE5iOEE&authuser=0) control counts (reference)

Please note, these files can also be downloaded using wget (sometimes these links do not work, for unclear reason):

```bash
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITUDZ0aXFBd2pDR0k -O M62_NA13019.tsv
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITWU9xTndtZE5iOEE -O M62_NA12878.tsv
```

A file with gene names and coordinates is required if clusters should be named using gene names. For human assemblies GRCh37 
(hg19) and GRCh38 (hg38), these file can be downloaded from [UCSC](http://hgdownload.soe.ucsc.edu/downloads.html):

* [refGene.txt.gz](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/refGene.txt.gz) assembly hg19 **(5 MB)**
* [refGene.txt.gz](http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/refGene.txt.gz) assembly hg38 **(5 MB)**

Clustered count files ready for processing with the final step are these:

* [M62_NA13019.clustered](https://drive.google.com/open?id=0BzLnl09R3GITYTduNDY4azJNZXM&authuser=0) 
* [M62_NA12878.clustered](https://drive.google.com/open?id=0BzLnl09R3GITWm1FS0duczVlejQ&authuser=0)

Please note, these files can also be downloaded using wget (sometimes these links do not work, for unclear reason):

```bash
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITYTduNDY4azJNZXM -O M62_NA13019.clustered
$ wget --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITWm1FS0duczVlejQ -O M62_NA12878.clustered
```

## Running

You can run all steps separately:

```bash
# extract counts
$ qgetcounts -i M62_NA13019.bam -a CNA902Y.bed > M62_NA13019.tsv
$ qgetcounts -i M62_NA12878.bam -a CNA902Y.bed > M62_NA12878.tsv

# cluster the counts
$ qcluster -i M62_NA13019.tsv [--names refGene.txt] > M62_NA13019.clustered
$ qcluster -i M62_NA12878.tsv [--names refGene.txt] > M62_NA12878.clustered

# call copy numbers
$ quandico --no-cluster \
   -s data=M62_NA13019.clustered \   # file with clustered counts
   -r data=M62_NA12878.clustered \   # file with clustered counts
   -s x=2 -s y=0 -r x=2 -r y=0   \   # sample (-s) and reference (-r) are female
   [--cp names=refGene.txt]          # optional for naming of clusters
```

Alternatively, all this can be done using one single command:

```bash
$ quandico -s map=M62_NA13019.bam -s x=2 -s y=0 \ # sample
           -r map=M62_NA12878.bam -r x=2 -r y=0 \ # reference
           -a CNA902Y.bed                       \ # amplicons
           -d results -b 13019_vs_12878         \ # output location and name
           [--cp names=refGene.txt]               # optional cluster names
```

## Example Output

In the folder 'ExampleOutput' the result files obtained by running the above mentioned steps with sample NA13019 and 
reference NA12878 from sequencing run M62 are presented. These are:

* [NA13019_overview.pdf](https://github.com/reineckef/quandico/raw/master/ExampleOutput/NA13019_overview.pdf) an overview for all clusters on one page
* [NA13019.pdf](https://github.com/reineckef/quandico/raw/master/ExampleOutput/NA13019.pdf) a detailed plot for every cluster
* [NA13019.vcf](https://github.com/reineckef/quandico/raw/master/ExampleOutput/NA13019.vcf) [VCF][vcf] of called variants
* [NA13019.csv](https://github.com/reineckef/quandico/raw/master/ExampleOutput/NA13019.csv) the called variants in tabular format

An image from one page (number 19) from the detailed PDF shows a loss (top) and a gain (bottom) event in that sample on 
chromsome X:

![Example page](/ExampleOutput/NA13019_p19.png)


## Help
The Perl command-line tools present a help and usage information when run without arguments (or by explicitly using `-h` 
or `--help`).

The `R` function provides help (after installation) by doing:

```R
R> library(quandico)
R> ?quandico
```

## Issues
Please report bugs to the author [Frank Reinecke][me] or file 
them on quandico's home [on GitHub](http://github.com/reineckef/quandico).


<!-- REFERENCES -->

[cnv]: http://en.wikipedia.org/wiki/Copy-number_variation "Copy Number Variation (Wikipedia)"
[cpan]: http://www.cpan.org "Comprehensive Perl Archive Network"
[cran]: http://cran.r-project.org "CRAN: Comprehensive R Archive Network"
[latest]: https://github.com/reineckef/quandico/releases/latest "the latest release"
[me]: mailto:frank.reinecke@qiagen.com
[perl]: http://www.perl.org "Perl Homepage"
[r]: http://www.r-project.org "R Project"
[samtools]: http://www.htslib.org "Samtools Homepage"
[strawberry]: http://www.strawberryperl.com "Strawberry Perl"
[vcf]: https://github.com/samtools/hts-specs "VCF Specifications"
[installer]: https://github.com/reineckef/quandico/raw/master/install.sh "install script"


PerlModule/Changes
commit 17c02d450c4908defbac44118713597d5acb3493
Merge: 636468a 9156bcc
Author: reineckef <mail@frank-reinecke.de>
Date:   Wed Sep 7 15:51:57 2016 +0200

    Merge branch 'develop'

commit 9156bcc240e2cf3168c3a127889f30c850f6f85a
Author: reineckef <mail@frank-reinecke.de>
Date:   Wed Sep 7 15:46:46 2016 +0200

    Updated version in DESCRIPTION and generated documentation files

 6 files changed, 11 insertions(+), 7 deletions(-)

commit ee017a286036b0b0f1859fd6db6a9707de8704aa
Author: reineckef <mail@frank-reinecke.de>
Date:   Wed Sep 7 15:43:02 2016 +0200

    Changed theme definitions to work with recent ggplot2 versions (tested
    with 2.1.0)

 1 files changed, 4 insertions(+), 3 deletions(-)

commit 7ea7d9d0e770db06c436343fc9a0e0bde23127c4
Author: reineckef <mail@frank-reinecke.de>
Date:   Wed Sep 7 15:35:13 2016 +0200

    Updated branch/version to v1.14 for PerlModule/lib/QUANDICO.pm

 1 files changed, 3 insertions(+), 2 deletions(-)

commit a3e4269488770f9697c61a1b0f8954c444ee77f1
Author: reineckef <mail@frank-reinecke.de>
Date:   Wed Sep 7 15:35:09 2016 +0200

    Set develop version to v1.14

 3 files changed, 55 insertions(+), 0 deletions(-)

commit 636468aeeabfeba533684863182ab7b010828f53
Author: reineckef <mail@frank-reinecke.de>
Date:   Wed Nov 26 08:50:09 2014 +0100

    Added comment line to automatically increment version number by a script

 1 files changed, 1 insertions(+), 1 deletions(-)

commit 5b049b7cfebbcb6e4fd71caa9a99feef59708eef
Merge: 54fb66c 2c3ed04
Author: reineckef <mail@frank-reinecke.de>
Date:   Mon Nov 24 18:20:21 2014 +0100

    Merge branch 'master' of github.com:reineckef/quandico

commit 2c3ed04a89c234a61986f6e9575b3a1259088cbd
Author: Frank Reinecke <frank.reinecke@qiagen.com>
Date:   Mon Nov 24 18:31:01 2014 +0100

    Fixed some typos in example code

 1 files changed, 2 insertions(+), 2 deletions(-)

commit 54fb66c16f12c53b341bcc807e79be6334a98cd5
Author: reineckef <mail@frank-reinecke.de>
Date:   Mon Nov 24 15:42:14 2014 +0100

    Use which to set default executables in install.sh

 1 files changed, 10 insertions(+), 7 deletions(-)

commit 80aa62cf7f6508defafff6d311a56bd3df3c8485
Author: reineckef <mail@frank-reinecke.de>
Date:   Mon Nov 24 15:39:26 2014 +0100

    Version change to 1.13 in documentation

 5 files changed, 8 insertions(+), 8 deletions(-)

commit c0bf042dcfc58a6acd7a476c97cd347c7902c9a8
Author: reineckef <mail@frank-reinecke.de>
Date:   Fri Nov 21 16:51:59 2014 +0100

    Set defaults to be more neutral, added 1.13 to README.md

 2 files changed, 8 insertions(+), 8 deletions(-)

commit f56a74501e795da38d4b044e976bee91e9247b18
Author: reineckef <mail@frank-reinecke.de>
Date:   Fri Nov 21 16:46:09 2014 +0100

    Set master version to 1.13

 1 files changed, 110 insertions(+), 0 deletions(-)

commit 96a0ae4a96b7eb39276545812759f9db2e632e85
Author: reineckef <mail@frank-reinecke.de>
Date:   Fri Nov 21 16:45:32 2014 +0100

    Added a description of install.sh to README, a lot of clean-up (removed
    /path/to in examples).

 2 files changed, 212 insertions(+), 79 deletions(-)

commit 0c5e774231b9068038b18f310f4517a2980a0788
Author: reineckef <mail@frank-reinecke.de>
Date:   Fri Nov 21 16:38:11 2014 +0100

    Set version in R to 1.13 to match the Perl package version

 2 files changed, 3 insertions(+), 3 deletions(-)

commit 2e499c0554dc99530c96c8fe50fe6452fdf051b2
Author: reineckef <mail@frank-reinecke.de>
Date:   Fri Nov 21 16:37:49 2014 +0100

    Removed testing dependencies (all required now), no high MakeMaker
    version needed any more.

 1 files changed, 6 insertions(+), 8 deletions(-)

commit 43dd4fd61f0876537de41379a55daaf4f94f0edd
Author: reineckef <mail@frank-reinecke.de>
Date:   Fri Nov 21 16:37:14 2014 +0100

    Changed output to always appear on STDOUT, so upstream tool usage is
    consistently "$ tool -arg > output.file"; increased version to 1.13

 4 files changed, 33 insertions(+), 40 deletions(-)

commit aa64a4fbec274443a55c4c2ce82f031cfb16c1e7
Author: reineckef <mail@frank-reinecke.de>
Date:   Thu Nov 20 19:23:03 2014 +0100

    Added 'install.sh' as a one-stop script to download and install
    everything.

 1 files changed, 149 insertions(+), 0 deletions(-)

commit 128e7484dab1e43b039b28d9938ba76f52aa62a6
Author: reineckef <mail@frank-reinecke.de>
Date:   Tue Nov 18 15:08:29 2014 +0100

    Fixed link to latest release

 1 files changed, 3 insertions(+), 2 deletions(-)

commit 63aca966b74c95b1aba6468b2f2bb28863ae84e8
Author: reineckef <mail@frank-reinecke.de>
Date:   Tue Nov 18 15:05:50 2014 +0100

    Added link to /releases/latest to download packages

 1 files changed, 3 insertions(+), 2 deletions(-)

commit 32327c062a5fe1ff7d301d41548c4af1c97146b2
Author: reineckef <mail@frank-reinecke.de>
Date:   Tue Nov 18 14:46:24 2014 +0100

    Added more documentations, many links and an example image

 2 files changed, 93 insertions(+), 38 deletions(-)

commit ac86e8cbe9646cfbba91ea19aee56f9a32fd5200
Author: reineckef <mail@frank-reinecke.de>
Date:   Tue Nov 18 14:26:52 2014 +0100

    Added ExampleOutput folder and content to show results from the
    demonstration run documented in README

 4 files changed, 130 insertions(+), 0 deletions(-)

commit ca10ee8a9e9f00cbd35d6144c95bde278a7b704e
Author: reineckef <mail@frank-reinecke.de>
Date:   Tue Nov 18 11:52:18 2014 +0100

    Added count data files and the corresponding amplicon coordinates (bed)
    in /Data

 3 files changed, 8915 insertions(+), 0 deletions(-)

commit 834b3f6986a43bc2bcbe03005fbe550a01465560
Author: reineckef <mail@frank-reinecke.de>
Date:   Mon Nov 17 17:24:47 2014 +0100

    Updated links

 5 files changed, 10 insertions(+), 10 deletions(-)

commit 8316affb7b472dfcc94146830a1c9daa064221bb
Author: fr02081975 <mail@frank-reinecke.de>
Date:   Mon Nov 17 17:18:36 2014 +0100

    Added both packages of version 1.12

 2 files changed, 0 insertions(+), 0 deletions(-)

commit e284b220c27044c82ab56072af25f7209822f6f4

PerlModule/Makefile.PL
use 5.008;
use strict;
use warnings;
use ExtUtils::MakeMaker;
WriteMakefile(
	NAME          => 'QUANDICO',
	AUTHOR        => q{Frank Reinecke <frank.reinecke@qiagen.com>},
	VERSION_FROM  => 'lib/QUANDICO.pm',
	ABSTRACT_FROM => 'lib/QUANDICO.pm',
	LICENSE       => 'gpl',
	EXE_FILES     => ['scripts/quandico', 'scripts/qcluster', 'scripts/qgetcounts'],

	#	PL_FILES  => {},
	PREREQ_PM => {
		'DBD::SQLite'               => 0,
		'DBI'                       => 1.622000,
		'Data::Dumper'              => 0,
		'DateTime'                  => 1.030000,
		'Env::Path'                 => 0,
		'File::Basename'            => 2.770000,
		'File::Copy'                => 2.140000,
		'File::Spec::Functions'     => 3.400000,
		'File::Temp'                => 0.230400,
		'File::chdir'               => 0.100400,
		'Getopt::Long::Descriptive' => 0.091000,
		'IPC::Run'                  => 0,
		'IPC::Cmd'                  => 0,
		'List::Util'                => 1.380000,
		'List::MoreUtils'           => 0,
		'Sort::Naturally'           => 1.020000,
		'version'                   => 0.770000,
		'Term::ANSIColor'           => 2.000000,
		'Test::More'                => 0,
		'Test::Script'              => 1
	},
	dist  => {COMPRESS => 'gzip -9f', SUFFIX => 'gz'},
	clean => {FILES    => 'QUANDICO-*'},
);


PerlModule/lib/QUANDICO.pm
package QUANDICO;
use 5.008;
use strict;
use warnings;
#
use Data::Dumper;
use DateTime;
use DBI;
use DBD::SQLite;
use Env::Path qw<PATH>;
use File::Basename qw<fileparse>;
use File::chdir;
use File::Copy qw<copy move>;
use File::Spec::Functions qw<rel2abs catfile>;
use File::Temp;
use Getopt::Long::Descriptive;
use List::Util qw<min max>;
use Sort::Naturally qw<nsort>;
use version 0.77;
our $VERSION = version->declare(v1.14); ## GIT_VERSION ## our $VERSION = version->declare(<%>);
our $BRANCH = 'master'; ## GIT_BRANCH ## our $BRANCH = '<%>';

# constants
use constant PAIRED     => 0x1;
use constant PROPERLY   => 0x2;
use constant UNMAPPED   => 0x4;
use constant MATENOTM   => 0x8;
use constant REVERSE    => 0x10;
use constant FIRST      => 0x40;
use constant SECOND     => 0x80;
use constant SPLIT      => 0x100;
use constant FAILED     => 0x200;
use constant DUPLICATE  => 0x400;
use constant SUPPLEMENT => 0x800;

# exporting
require Exporter;
our @ISA    = qw(Exporter);
our @EXPORT = qw<
describe_options
fileparse copy move rel2abs catfile
min max nsort
Dumper
$CWD abs_path PATH
show_version
>;    # symbols to export always
our @EXPORT_OK = qw<PAIRED PROPERLY UNMAPPED MATENOTM REVERSE
FIRST SECOND SPLIT FAILED DUPLICATE SUPPLEMENT>;
our %EXPORT_TAGS = (
	FLAGS => [
		qw<PAIRED PROPERLY UNMAPPED MATENOTM REVERSE FIRST
		SECOND SPLIT FAILED DUPLICATE SUPPLEMENT >
	]
);

=head1 NAME

QUANDICO - Package with helper scripts for the R application 'quandico'

=head1 VERSION

Version 1.12

=cut

=head1 SYNOPSIS

This module holds documentation and some commonly used routines for the script 
distributed along with this package. This modules is not meant to be of any 
general use instead of providing a container for documentation and provide 
the means to package scripts for easy installation (with dependencies).

=over 3

=item * quandico - driver script to prepare and run the R application

=item * qcluster - script to group neighboring amplicon count data into groups

=item * qgetcounts - script to extract read depth for amplicon coordinates from SAM/BAM

=back

Please check the documentation of these script for detailed help. Generally, running 
these with "--help" will present usage information. Generally, quandico is the main 
driver which is also able to run the other two helper script for you.

=head1 EXPORT

Some functions of commonly used external modules are exported to the helper scripts. 
Please check the code if you need to know the details.

SAM flags can be exported by request:

  use QUANDICO qw<:FLAGS>;
  


=head1 SUBROUTINES/METHODS

Subroutines contained in here are also only used by the scripts mentioned above, and 
these are not meant to be of any general use. 


=head2 show_version()

This prints the version number and exits.

=cut

sub show_version {
	my $name = $0;
	$name =~ s!^\.\/!!;
	printf "This is %s %s %s\nUse flags -h or --help for help.\n", $name, $VERSION->stringify, $BRANCH ? qq~($BRANCH)~ : '';
	exit;
}

=head2 abs_path() 

This will create an absolute path via rel2abs(), but always uses slashes and never 
backslash, even on Windows. The path is meant to be used inside R.

=cut

sub abs_path {
	my $abs = rel2abs(shift);
	$abs =~ tr/\\/\//;
	return $abs;
}

=head1 AUTHOR

Frank Reinecke, C<< <frank.reinecke at qiagen.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<frank.reinecke at qiagen.com>, 
or through the website at L<https://github.com/fr02081975/quandico>. 


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc QUANDICO


You can also look for information at:

=over 4

=item * GitHub (report bugs there)

L<https://github.com/fr02081975/quandico>


=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 Frank Reinecke.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 3 dated June, 2007 or at your option
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

A copy of the GNU General Public License is available in the source tree;
if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


=cut

1;    # End of QUANDICO





install.sh
#!/bin/sh
#
###############################################################################
#
# This script is written to download and install the R package 'quandico'
# and the Perl module 'QUANDICO' with required dependencies and test data.
# 
# Obviously, this does only work on Unix/Linux systems and should be seen
# as a shortcut for all manual steps described in the 'README.md' file.
#
# *** This is alpha software and not guaranteed to work! ***
# 
# Adjust these to match your needs. You may want to skip the BAM file download
# if you are limited in disk space (this needs ~500 MB of space) or your
# download speed is low.
#
# *** CONFIGURATION *** // START
#
# executables
REXE=$(which R);            # point to the R installation to use
PERL=$(which perl);         # point to the Perl installation to use
SAMTOOLS=$(which samtools); # specify which samtools to use (must be >= version 1.1)
#
# settings
BIGDATA=0; # set to 1 if you would like to download BAM files (~474 MB)
REFGENE=1; # set to 0 if you would like to skip refGene.txt download (~15 MB)
TESTRUN=1; # set to 0 if you would like to skip the test of 'quandico'
CPANM=0;   # set to 1 if you prefer App::cpanminus over classic 'make'
#
# *** CONFIGURATION *** // END
###############################################################################
#
# Some Shortcuts for Output Formatting
#
function action() { echo -e "\n\033[1;7m$1\033[0m\n" ; }
function skip()   { echo -e "\n\033[2;7m$1\033[0m\n" ; }
function good()   { echo -e "\033[32m$1\033[0m" ; }
function bad()    { echo -e "\033[31m$1\033[0m" ; }
function bail()   { echo -e "\033[31;7m$1\033[0m" ; exit; }

#
###############################################################################
#
# STEP 0: Basic prereqs
#
# Need R
action "[ PREREQ ] Testing for R, Perl and Samtools ... "
if [ -x $REXE ]; then 
  good "$REXE is executable and will be used"
else
  bail "$REXE is not executable"
fi
# Need Perl
if [ -x $PERL ]; then 
  good "$PERL is executable and will be used"
else
  bail "$PERL is not executable"
fi

# Need samtools for BAM
if [ -x $SAMTOOLS ]; then 
  STV=$($SAMTOOLS --version | perl -ne "if (/samtools ([\d\.]*)/gsm) { \$v = \$1 >= 1.1 ? 'ok: '. \$1 : 'old: ' . \$1 }; END { print \$v }")
  if [[ $STV == ok* ]]; then
    good "$SAMTOOLS is executable, the version is $STV (>= 1.1)."
  else
    bad "
Cannot run on BAM files without samtools version >= 1.1!
You have samtools $STV. Please consider upgrading to the most recent version 
of 'samtools' from http://www.htslib.org to use BAM files as input.
"

    if (( $BIGDATA + $TESTRUN == 2 )); then
      bail "Cannot run on BAM (BIGDATA=$BIGDATA, TESTRUN=$TESTRUN) files without samtools!"
    fi
  fi

else

  bad "
Cannot use BAM files without samtools (version >= 1.1)!
Please consider downloading the most recent version of 'samtools'
from http://www.htslib.org to use BAM files as input.
"

  if (( $BIGDATA + $TESTRUN == 2 )); then
    bail "Cannot run on BAM (BIGDATA=$BIGDATA, TESTRUN=$TESTRUN) files without samtools!"
  fi
fi

if [ -e latest ]; then 
  action "[  CLEAN  ] Removing previous 'latest' file ... "
  rm latest
fi

###############################################################################
#
# STEP 1: Find the links for the latest release
#

action "[ STEP  1  ] Looking for the latest version ... "
wget -nv https://github.com/reineckef/quandico/releases/latest -O latest

###############################################################################
#
# STEP 2: Downloads
#
RPACK=$(grep -Po "(releases/download.*.quandico.*\.tar\.gz)" < latest)
action "[ STEP  2a ] Downloading the R package: $RPACK ... "
wget -nc -nv --no-check-certificate https://github.com/reineckef/quandico/$RPACK

PERLM=$(grep -Po "(releases/download.*.QUANDICO.*\.tar\.gz)" < latest)
action "[ STEP  2b ] Downloading the Perl module: $PERLM ... "
wget -nc -nv --no-check-certificate https://github.com/reineckef/quandico/$PERLM

action "[ STEP  2c ] Downloading example data ... "
wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITQjBUZUFVcy1BNFk -O CNA902Y.bed
wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITUDZ0aXFBd2pDR0k -O M62_NA13019.tsv
wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITWU9xTndtZE5iOEE -O M62_NA12878.tsv
wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITYTduNDY4azJNZXM -O M62_NA13019.clustered
wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITWm1FS0duczVlejQ -O M62_NA12878.clustered

if [ $BIGDATA == 1 ]; then
  action "[ STEP  2d ] Downloading BAM files (474 MB), please wait ... "
  wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITMzNyakhveTh3UVE -O M62_NA13019.bam
  wget -nc -nv --no-check-certificate https://googledrive.com/host/0BzLnl09R3GITSnU1TlVRSjRXRHM -O M62_NA12878.bam
else
  skip "[ STEP  2d ] Skipping BAM files "
fi


if [ $REFGENE == 1 ]; then
  action "[ STEP  2e ] Downloading refGene ... "
  wget -nc -nv -O- http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/refGene.txt.gz | gunzip > refGene.txt
  WITHREF="--cp names=refGene.txt"
else
  skip "[ STEP  2e ] Skipping refGene.txt "
  WITHREF=""
fi

###############################################################################
#
# STEP 3: Install R dependencies
#
action "[ STEP  3  ] Installing R dependencies..."
$REXE --slave --vanilla <<END_OF_R

# options(warn=-1)
# update.packages(repos="http://cran.rstudio.com/", ask=FALSE) # this will update ALL packages

for (pkg in c("MASS", "grid", "ggplot2", "gridExtra", "naturalsort", "scales")) {
  check <- require( pkg , character.only = TRUE )
  if (check) {
    message(cat(pkg, 'is already installed', sep=" "))
    update.packages(pkg, repos="http://cran.rstudio.com/", ask=FALSE)
  } else {
    message(cat("installing", pkg, "...", sep=" "))
    install.packages(pkg, repos="http://cran.rstudio.com/")
  }
}
q()

END_OF_R

echo "OK."

###############################################################################
#
# STEP 4: Install R package
#
RPACK=$(perl -e "@L = split(/\//, '$RPACK'); print \$L[-1];")
action "[ STEP  4  ] Installing 'quandico' R package '$RPACK' ... "
$REXE CMD INSTALL $RPACK
echo "OK."

###############################################################################
#
# STEP 5: Install Perl Module
#
PERLM=$(perl -e "@L = split(/\//, '$PERLM'); print \$L[-1];")
PURE=$(perl -e "print \$1 if '$PERLM' =~ /^(.*?)\.tar\.gz/;")

action "[ STEP  5  ] Installing Perl Module '$PERLM' ... "

if [ $CPANM == 1 ]; then
  echo "Installing Perl Module '$PURE' using cpanm (App::cpanminus) ..."
  cpan App::cpanminus && cpanm -v $PERLM
  echo "OK."
  echo ""
else
  echo "Installing Perl Module '$PURE' the classic way ..."
  tar xvfz $PERLM
  cd $PURE
  perl Makefile.PL
  make
  make test
  make install
  cd ..
  echo "OK."
  echo ""
fi

###############################################################################
#
# STEP 6: Run tests
#

if [ $TESTRUN != 1 ]; then
  skip "[ STEP  6  ] Test run not requested, skipping. "

else
  action "[ STEP  6  ] Testing the installation on demo data $WITHREF ... "

  if [ $BIGDATA == 1 ]; then
    echo "Running 'quandico' on BIG test data, this can take up to ten minutes to complete ..."
    echo "quandico -A CNA902Y.bed -X samtools=$SAMTOOLS --rexe $REXE -s map=M62_NA13019.bam -r map=M62_NA12878.bam -s x=2 -s y=0 -r x=2 -r y=0 $WITHREF"
    time quandico -A CNA902Y.bed -X samtools=$SAMTOOLS --rexe $REXE -s map=M62_NA13019.bam -r map=M62_NA12878.bam -s x=2 -s y=0 -r x=2 -r y=0 $WITHREF
  else
    echo "Running 'quandico' on extracted test data, this should take less than one minute ..."
    echo "quandico --rexe $REXE -s data=M62_NA13019.tsv -r data=M62_NA12878.tsv -s x=2 -s y=0 -r x=2 -r y=0 $WITHREF"
    time quandico --rexe $REXE -s data=M62_NA13019.tsv -r data=M62_NA12878.tsv -s x=2 -s y=0 -r x=2 -r y=0 $WITHREF
  fi

fi

action "[  RESULTS  ] You should see some files named 'sample_vs_reference.[ext]' "
ls -hltr sample_vs_*

good "

The R package and the Perl Module (both with dependencies)
should now be installed. Please check output for errors.

"

